README.md

<a href="https://www.buymeacoffee.com/playlinesdev" target="_blank"><img src="https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png" alt="Buy Me A Coffee" style="height: auto !important;width: auto !important;" ></a>

# pl_grid

The PlGrid is a grid view for showing data in a table with a sticky header and pagination components.

## Nice example of the customization possibilities

<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_customizations.gif?raw=true"/>

<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_nice.gif?raw=true"/>

The code for the above result
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_nice_code.png?raw=true"/>

## Getting Started

To add the grid on your project, install it by adding the pl_grid package dependency to your pubspec.yaml file, then import the PlGrid class like so: 
```dart
import 'package:pl_grid/pl_grid.dart';
````
Add the widget like in the example down below:

```dart
PlGrid(
    headerColumns: ['Id', 'Name', 'Age'],
    data: [
        [1, 'Bruno', 34],
        [2, 'Lindsey', 39],
        [3, 'Roberto', 18],
        [4, 'Yasmin', 22]
    ],
    columnWidthsPercentages: [15, 70, 15],
    curPage: 1,
    maxPages: 4,
    width: 300,
    height: 200,
)
```

and it will produce the following:

<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample1.png?raw=true"/>

One can also use other parameters for example paginationItemClick that is a function that receives the index of the page touched/clicked to execute any routine such as call api with next page results

The property asCard is set as default to true, but could be set to false, producing the following:

<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample1.png?raw=true"/>

<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample2.png?raw=true"/>
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_web.png?raw=true"/>


A complex example to show the many features that can be achieved with this package

```dart
Scaffold(
      backgroundColor: Colors.black38,
      body: Padding(
        padding: const EdgeInsets.only(top: 80, left: 80),
        child: Container(
          child: PlGrid(
            internalGrid: true,
            width: 480,
            height: 400,
            asCard: true,
            invertZebra: true,
            searchBarHeight: 30,
            searchBarTextAlign: TextAlign.center,
            showSearchBar: true,
            paginationItemClick: (pageClicked) {
              print('Call the api with page attribute');
            },
            noContentWidget: Center(
              child: Container(
                width: 30,
                height: 30,
                child: CircularProgressIndicator(),
              ),
            ),
            onSearch: (searchTextTyped) {
              print(searchTextTyped);
            },
            headerCellRenderer: (cell, content) => cell == 2
                ? FlatButton(
                    onPressed: () {},
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                      children: <Widget>[
                        Text(content),
                        Icon(Icons.settings_applications)
                      ],
                    ),
                  )
                : null,
            headerAlignmentByCells: (i) =>
                i % 2 == 0 ? Alignment.topRight : Alignment.bottomRight,
            asCardPadding: PlGrid.baseAsCardPadding.copyWith(left: 35),
            headerCellsColor: (i) => Colors.blueGrey[50],
            heightByRow: (row) => row == 2 ? 40 : null,
            rowsCellRenderer: (row, cell, content) => row == 2 && cell == 2
                ? Icon(
                    Icons.sentiment_satisfied,
                    color: Colors.redAccent,
                  )
                : null,
            zebraStyle: PlGrid.baseZebraStyle.copyWith(color: Colors.blueGrey),
            rowCellsPadding: PlGrid.baseRowCellsPadding.copyWith(right: 10),
            alignmentByRow: (row, cell) => row == 2 && cell == 2
                ? Alignment(-0.85, -1)
                : Alignment.bottomRight,
            headerStyle: PlGrid.baseHeaderStyle.copyWith(fontSize: 13),
            columnWidthsPercentages: <double>[25, 25, 25, 25],
            headerColumns: ['Id', 'Name', 'Age', 'aa'],
            maxPages: 3,
            curPage: 1,
            data: [
              [1, 'Person A', 9, 'abc'],
              [2, 'Person B', 18, 'abc'],
              [3, 'Person C', 75, 'abc'],
              [4, 'Person D', 31, 'abc'],
              [5, 'Person E', 26, 'abc'],
              [6, 'Person F', 52, 'abc'],
              [7, 'Person G', 13, 'abc'],
              [8, 'Person H', 24, 'abc'],
            ],
          ),
        ),
      ),
    );
```

<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_complex.png?raw=true"/>
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_complex1.png?raw=true"/>


CHANGELOG.md
## [0.0.1] - TODO: Add release date.

* TODO: Describe initial release.

## [0.0.1+1] - Marks some parameters as @required

* Some parameters that are mandatory were annotated accordingly

## [0.0.0+2] - Adds image to documentation

Other documentation improvements
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample1.png?raw=true"/>

## [0.0.0+4] - Adds MIT License info

## [0.0.1+0] - Adds lots of customization possibilities to have control of each cell
Makes it also possible to render Widgets on any cell
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample4.png?raw=true"/>

## [0.0.3+0] - Adds search bar and many possible customizations

Below is a complex example of what is possible to achieve
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_complex.png?raw=true"/>

## [0.0.3+2] Improvements on documentation

## [0.0.4] Awesome improvements on documentation
Now every single parameter in the gigantic constructor is very well documented, sometimes with dart code example. To use the most of the widget.

## [0.1.0] Breaking changes
Converts the widget from a stateless widget to a stateful in order to control it better. Documentation also
improved

## [0.2.0] Many features
Added many features on onSearch event (search interval and run condition), onPaginationItemClick event. Also the possibility to set a ThemeData for the whole widget including it's container card

example:
<img src="https://github.com/playlinesdev/pl_grid/blob/master/sample_nice.gif?raw=true"/>

## [0.2.1] Improved customizations
Many customizations possibilities such as the possibility of pass different TextsStyles for the pagination item buttons, the possibility
to have asynchronous data and much more

test/pl_grid_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:pl_grid/pl_grid.dart';

PlGrid get _baseWidget => PlGrid(
      curPage: 1,
      maxPages: 10,
      headerColumns: ['Id', 'Name', 'Age'],
      data: [
        [1, 'Bruno', 34],
        [2, 'Lindsey', 39],
        [3, 'Roberto', 18],
        [4, 'Yasmin', 22]
      ],
      width: 300,
      height: 400,
      columnWidthsPercentages: <double>[15, 70, 15],
    );

MaterialApp get _baseApp => MaterialApp(
      home: Scaffold(
        body: Container(
          width: 480,
          height: 800,
          child: Center(
            child: _baseWidget,
          ),
        ),
      ),
    );

void main() {
  testWidgets('It should not take any exception', (tester) async {
    await tester.pumpWidget(_baseApp);

    expect(tester.takeException(), null);
  });
  testWidgets('It should find a widget for calling the page 10',
      (tester) async {
    await tester.pumpWidget(_baseApp);
    expect(find.text('10'), findsOneWidget);
  });
}


example/pl_grid_example.dart
import 'package:flutter/material.dart';
import 'package:pl_grid/pl_grid.dart';

void main() => runApp(
      MaterialApp(
        home: Scaffold(
          body: PlGrid(
            headerColumns: ['Id', 'Name', 'Age'],
            data: [
              [1, 'Bruno', 34],
              [2, 'Lindsey', 39],
              [3, 'Roberto', 18],
              [4, 'Yasmin', 22]
            ],
            onPaginationItemClick: (i) async {
              await Future.delayed(Duration(seconds: 3));
              return [];
            },
            paginationDynamicStyle: (page) => TextStyle(color: Colors.blue),
            dataAsync: () async {
              await Future.delayed(Duration());
              return [];
            },
            columnWidthsPercentages: <double>[15, 70, 15],
            curPage: 1,
            maxPages: 4,
          ),
        ),
      ),
    );


example/pl_grid_example2.dart
import 'package:flutter/material.dart';
import 'package:pl_grid/pl_grid.dart';

void main() => runApp(
      MaterialApp(
        home: Scaffold(
          body: grid,
        ),
      ),
    );

get grid => PlGrid(
      headerColumns: [
        'Id',
        FlatButton.icon(
            onPressed: () {},
            icon: Icon(
              Icons.add_box,
              size: 8,
            ),
            label: Text(
              'Product',
              style: TextStyle(fontSize: 13),
            )),
        'Price',
        'Stock',
        'Action'
      ],
      headerStyle: TextStyle(
        fontSize: 12,
        fontWeight: FontWeight.bold,
      ),
      rowsCellRenderer: (row, cell, content) {
        if (row == 2 && cell == 2) return Icon(Icons.accessibility_new);
        return null;
      },
      heightByRow: (i) => 35,
      rowsStyle: TextStyle(fontSize: 12),
      width: 320,
      height: 200,
      asCardPadding: EdgeInsets.all(10),
      onPaginationItemClick: (i) {
        print('touched pagination page $i button');
      },
      data: data,
      headerHeight: 25,
      alignmentByRow: (row, cell) => Alignment.topCenter,
      headerAlignmentByCells: (cell) => Alignment.center,
      columnWidthsPercentages: <double>[8, 32, 20, 20, 20],
      curPage: 1,
      maxPages: 2,
    );

get data => [
      [
        1,
        'Pencil',
        1.5,
        30,
        FlatButton.icon(
            onPressed: () {},
            icon: Icon(
              Icons.access_alarm,
              size: 10,
            ),
            label: Text(
              'add',
              style: TextStyle(fontSize: 8),
            ))
      ],
      [1, 'Pen', 2.5, 25, 1],
      [1, 'Notebook', 8.5, 20, 1],
      [1, 'Note', 8.5, 20, 1],
      [1, 'Pen', 8.5, 20, 1],
    ];


example/pl_grid_complex_example.dart
import 'package:flutter/material.dart';
import 'package:pl_grid/pl_grid.dart';

class PlGridExample extends StatefulWidget {
  @override
  _PlGridExampleState createState() => _PlGridExampleState();
}

class _PlGridExampleState extends State<PlGridExample> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      backgroundColor: Colors.black38,
      body: Padding(
        padding: const EdgeInsets.only(top: 80, left: 80),
        child: Container(
          child: PlGrid(
            internalGrid: true,
            width: 480,
            height: 400,
            asCard: true,
            invertZebra: true,
            searchBarHeight: 30,
            searchBarTextAlign: TextAlign.center,
            showSearchBar: true,
            onPaginationItemClick: (pageClicked) {
              print('Call the api with page attribute');
            },
            noContentWidget: Center(
              child: Container(
                width: 30,
                height: 30,
                child: CircularProgressIndicator(),
              ),
            ),
            onSearch: (searchTextTyped) {
              print(searchTextTyped);
            },
            headerCellRenderer: (cell, content) => cell == 2
                ? FlatButton(
                    onPressed: () {},
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                      children: <Widget>[
                        Text(content),
                        Icon(Icons.settings_applications)
                      ],
                    ),
                  )
                : null,
            headerAlignmentByCells: (i) =>
                i % 2 == 0 ? Alignment.topRight : Alignment.bottomRight,
            asCardPadding: PlGrid.baseAsCardPadding.copyWith(left: 35),
            headerCellsColor: (i) => Colors.blueGrey[50],
            heightByRow: (row) => row == 2 ? 40 : null,
            rowsCellRenderer: (row, cell, content) => row == 2 && cell == 2
                ? Icon(
                    Icons.sentiment_satisfied,
                    color: Colors.redAccent,
                  )
                : null,
            zebraStyle: PlGrid.baseZebraStyle.copyWith(color: Colors.blueGrey),
            rowCellsPadding: PlGrid.baseRowCellsPadding.copyWith(right: 10),
            alignmentByRow: (row, cell) => row == 2 && cell == 2
                ? Alignment(-0.85, -1)
                : Alignment.bottomRight,
            headerStyle: PlGrid.baseHeaderStyle.copyWith(fontSize: 13),
            columnWidthsPercentages: <double>[25, 25, 25, 25],
            headerColumns: ['Id', 'Name', 'Age', 'aa'],
            maxPages: 3,
            curPage: 1,
            data: [
              [1, 'Person A', 9, 'abc'],
              [2, 'Person B', 18, 'abc'],
              [3, 'Person C', 75, 'abc'],
              [4, 'Person D', 31, 'abc'],
              [5, 'Person E', 26, 'abc'],
              [6, 'Person F', 52, 'abc'],
              [7, 'Person G', 13, 'abc'],
              [8, 'Person H', 24, 'abc'],
            ],
          ),
        ),
      ),
    );
  }
}
